cirq-google/cirq_google/transformers/analytical_decompositions/two_qubit_to_sycamore.py

```diff
@@ -0,0 +1,464 @@
+# Copyright 2022 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Utility methods for decomposing two-qubit unitaries into Sycamore gates."""
+
+from typing import Iterator, List, Optional
+
+import itertools
+import math
+import numpy as np
+import scipy.linalg
+
+import cirq
+from cirq_google import ops
+
+UNITARY_ZZ = np.kron(cirq.unitary(cirq.Z), cirq.unitary(cirq.Z))
+PAULI_OPS = [
+    np.eye(2),
+    cirq.unitary(cirq.X),
+    cirq.unitary(cirq.Y),
+    cirq.unitary(cirq.Z),
+]
+
+
+def _decompose_arbitrary_into_syc_tabulation(
+    op: cirq.Operation, tabulation: cirq.TwoQubitGateTabulation
+) -> cirq.OP_TREE:
+    """Synthesize an arbitrary 2 qubit operation to a Sycamore operation using the given Tabulation.
+
+    Args:
+        op: Operation to decompose.
+        tabulation: A `cirq.TwoQubitGateTabulation` for the Sycamore gate.
+
+    Yields:
+        A `cirq.OP_TREE` that performs the given operation using Sycamore operations.
+    """
+    qubit_a, qubit_b = op.qubits
+    result = tabulation.compile_two_qubit_gate(cirq.unitary(op))
+    local_gates = result.local_unitaries
+    for i, (gate_a, gate_b) in enumerate(local_gates):
+        yield from _phased_x_z_ops(gate_a, qubit_a)
+        yield from _phased_x_z_ops(gate_b, qubit_b)
+        if i != len(local_gates) - 1:
+            yield ops.SYC.on(qubit_a, qubit_b)
+
+
+def two_qubit_matrix_to_sycamore_operations(
+    q0: cirq.Qid,
+    q1: cirq.Qid,
+    mat: np.ndarray,
+    *,
+    atol: float = 1e-8,
+    clean_operations: bool = True,
+) -> cirq.OP_TREE:
+    """Decomposes a two-qubit unitary operation into `cirq_google.SYC` + single qubit rotations.
+
+    The analytical decomposition first Synthesizes the given operation using `cirq.CZPowGate` +
+    single qubit rotations and then decomposes each `cirq.CZPowGate` into `cirq_google.SYC` +
+    single qubit rotations using `cirq_google.known_2q_op_to_sycamore_operations`.
```

Comment:
Worst case this means 8 sycs for a given matrix ?

Reply:
Worst case should be 6 Syc's, since matrix --> CZs is bounded by 3 CZs and then we do 2 Syc's for every CZPow. I've added the bound of 6 in tests and added test cases for 10 random 2qubit unitary matrices to assert this.